### PR TITLE
Pyproject toml builds dbt

### LINKED ISSUE #5696: [Spike]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

In order to [switch to using `pyproject.toml`](#4674) we need to spike a few impact/feasibility things:
* Will this break things in `dbt-cloud`?  I'm specifically concerned about our docker images since I know they're a bit fragile install wise.
* Are all of the features we use in `setup.py` supported in `pyproject.toml`?  Our setuptools setup isn't terribly complex, but there's some uncertainty here (esp in adapter repos re the version checking regex stuff :grimace:)

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by davidbloss**
resolves #5696 and #4674 and #4446

### Description

This update to `pyproject.toml` allows pip to install dbt-core with `poetry` in place of `setuptools`. dbt adapters, other dependent codebases, core devs, and users should not notice any changes from this update. 

#### Notes
- pip serves as the "frontend" of the build process while both setuptools and poetry run on the "backend"
- [Installing dbt with pip](https://docs.getdbt.com/dbt-cli/install/pip) should behave the same between setuptools and poetry
- poetry is automagically managed by pip -- devs and users do not need to separately install it for purposes of installing dbt-core. pip does not download poetry for later use unless specifically told to
- `setup.py` and `MANIFEST.in` still remain but all their logic has been consolidated into pyproject.toml
- Either poetry or setuptools can be used at this state 
- `poetry.lock` provides a deterministic, version controlled option for dependency management

#### Mapping old keys to new keys

setup.py | pyproject.toml
--- | ---
long_description | readme
long_description_content_type |  readme
author | authors
author_email | authors
url | repository
entry_points | [tool.poetry.scripts]
- minimal python version enforced by `python = "<version>"` under `[tool.poetry.dependencies]` in place of `if sys.version_info ...` in setup.py

#### Dropped entries
`test_suite="test"` see [deprecated setuptools feature](https://setuptools.pypa.io/en/latest/deprecated/changed_keywords.html#new-and-changed-setup-keywords)
`zip_safe=False` see [other deprecated setuptools feature](https://setuptools.pypa.io/en/latest/deprecated/zip_safe.html#understanding-the-zip-safe-flag)

#### Added entries
`license = "Apache-2.0"`
`homepage = "https://getdbt.com"`
`documentation = "https://docs.getdbt.com/"`
Various links in `[tool.poetry.urls]`

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -20,6 +20,11 @@ repos:
     exclude_types:
       - "markdown"
   - id: check-case-conflict
+- repo: https://github.com/python-poetry/poetry
+  rev: 1.2.2
+  hooks:
+    - id: poetry-check
+    - id: poetry-lock
```
Comment (max-sixty) [inline .pre-commit-config.yaml:27]: Not a reviewer but — this is going to update the lock file on every run — is that what we want?

**Comment by leahwicz** [comment]
Hi @davidbloss! Thanks so much for opening this contribution up! We currently are a couple weeks away from scheduling work to revamp our package and release process for `dbt-core` and the adapters. Part of that work will involve looking at investing more in `pyproject.toml` like you are doing here. I would ask if you could please 🙏  be patient with us reviewing this PR until we get to that work in case it has any impact on our plans. I would hate to merge this and then immediately change something again depending on the work .
I actually would be interested in hearing more from you on if you have other suggestions or things you would like to see us do with changes around setup, packaging, etc. that we might be able to include in the scope of this upcoming work. It's still being planned out and designed so any feedback on our current setup would be appreciated!

**Comment by jtcohen6** [comment]
I hadn't seen this a month ago — very cool!

Does it simplify/unblock this work for us to first move `dbt-postgres` out into its own repository? (#6189)

We have been building our Docker images from specific release branches, rather than from the dist wheels built from those release branches. It does seem _feasible_ that we could always build a wheel first. Not sure how tricky this is in practice / for our release process.

**Comment by davidbloss** [comment]
Hey all, I am closing this since Poetry is IMO not the best path forward. Instead I will be reopening a similar branch that makes use of [hatch](https://hatch.pypa.io/latest/) - much better for CI and testing purposes

**Pull request closed without merging**